Add some functions from IOverlayAppletProxy and hid:sys

exelix11 opened the pull request:
I'm working on a custom overlay applet replacement, this implements some calls that provide functions specific to this system applet. All the functions in the headers are tested and working.

yellows8 commented on nx/include/switch/services/applet.h:
```diff
@@ -468,3 +468,18 @@ u32 appletGetPerformanceMode(void);
 Result appletSetFocusHandlingMode(AppletFocusHandlingMode mode);
+
+/**
+ * @brief Stops forwaring the input to the foreground app. Works only in the Overlay applet context
```
*forwarding

yellows8 replied:
periods

yellows8 replied:
Description seems completely different from cmd name?

exelix11 replied:
Yes, i couldn't understand the meaning at first but that's the only effect it has and works consistently to lock/unlock the input, makes sense if you think about it like "start displaying the quick action menu"

yellows8 commented on nx/include/switch/services/hidsys.h:
```diff
@@ -0,0 +1,20 @@
+#include "../kernel/event.h"
+#include "../services/sm.h"
+
+Result hidSysInitialize(void);
```
hidsys, and remove following newline.

yellows8 commented on nx/include/switch/services/hidsys.h:
```diff
@@ -0,0 +1,20 @@
+Result hidSysEnableAppletToGetInput(bool enable);
+
+/**
+* @brief Returns an event that fires when the home button is pressed, this will prevent the home menu from opening when the button is pressed.
```
"this will prevent ..." Was this really intended for this cmd?

exelix11 replied:
I think so, libappletRequestHomeMenu still works as expected and when closing the event the button starts behaving normally again.

yellows8 commented on nx/source/services/applet.c:
```diff
@@ -314,6 +316,8 @@ void appletExit(void)
             }
         }
+		
+		eventClose(&HomeButtonReaderLockAccessorEvent);
```
tabs->spaces

yellows8 commented on nx/source/services/applet.c:
```diff
@@ -2315,3 +2319,38 @@ bool appletMainLoop(void) {
 }
+
+Result appletBeginToWatchShortHomeButtonMessage(void) {
+	if (__nx_applet_type != AppletType_OverlayApplet)
```
whitespace

yellows8 commented on nx/source/services/applet.c:
```diff
@@ -2315,3 +2319,38 @@ bool appletMainLoop(void) {
+	
+	Service ILockAccessor = {0};
+	Result rc = _appletGetSession(&g_appletICommonStateGetter, &ILockAccessor, 30);
+	if (!R_SUCCEEDED(rc))
```
R_FAILED

yellows8 commented on nx/source/services/hidsys.c:
```diff
@@ -0,0 +1,215 @@
+static u64 g_hidSysAppletResourceUserId = 0;
+
+Result hidSysInitialize(void) {
+	atomicIncrement64(&g_hidSysRefCnt);
```
whitespace again

yellows8 commented on nx/source/services/hidsys.c:
```diff
@@ -0,0 +1,215 @@
+    IpcCommand c;
+    ipcInitialize(&c);
+
+    struct CmdStruct{
```
Update to match rest of libnx.

yellows8 commented on nx/source/services/hidsys.c:
```diff
@@ -0,0 +1,215 @@
+
+static Event g_hidSysHomeEvent = {0};
+static Event g_hidSysCaptureEvent = {0};
+static Event g_hidSysSleeptEvent = {0};
```
*g_hidSysSleepEvent

yellows8 commented:
Also document events' auto-clear in the comments.

exelix11 commented:
This commit should address all the issues.
While testing i also noticed that the overlay applet receives a 0x15 message to signal the long home button pression (that doesn't fire the two events I added), in order to handle that i added the [UnknownMessageHandler](https://github.com/switchbrew/libnx/pull/242/commits/fb31b8f18467ddeada3ac46c55e74417679b0dbd#diff-f02a6bfca6491f39c91f00ce31e27be3R22) weak function.

misson20000 commented:
There are a *lot* more messages that am has code for sending than what's documented on the wiki. You may want to come up with a better scheme for applet message handling.

exelix11 commented:
Unfortunately i couldn't think of any better way to signal unknown messages to the running app.
Initially I thought about implementing a mainLoop function in my app but as all service handles in libnx are static I would have to init them by myself and so I wouldn't be able to use most libnx wrappers. Another option would be to handle the messages in the lib which seems overkill for case-specific messages like this.
That's why i decided to add a "callback" function which allows handling messages for specific purposes and isn't invasive/breaking, i understand that it may not follow your design principles and if so i will remove it from this pr.
Can you link the wiki documentation you mentioned ? I couldn't find anything related to am messages, the way i'm figuring out the ones for overlay is by logging the unknown ones i get.

misson20000 commented:
They're on the applet manager page under the heading "AppletMessage". I'd link but I'm on mobile.
Also, worth noting: I'm not a libnx maintainer. If yellows8 & co think it's ok, that's their decision.

exelix11 commented:
Maybe i'm blind cause i still can't find it.... are we both talking about the [switchbrew page](https://switchbrew.org/wiki/Applet_Manager_services) ?
Anyway i'll wait for a maintainer answer

yellows8 commented:
https://switchbrew.org/wiki/Applet_Manager_services#NotificationMessage
See also the info immediately below this table: https://switchbrew.org/wiki/Applet_Manager_services#ICommonStateGetter

Yeah that UnknownMessageHandler isn't great. Have you looked at the official overlay-applet code which handles messages?

exelix11 commented:
Oh it was at the very bottom of the page, thanks.
Tbh i'm not very skilled with RE and haven't found the message handling code, nor was i looking for it, while reading through libnx code i thought that there could have been unhandled messages.
Anyway shall i remove the UnknownMessageHandler ? Do you have any suggestion for a better implementation ? 
Needless to say in the case i removed it from the PR i will keep in my fork as libnx doesn't have any other way to let the app handle messages.

yellows8 commented:
Maybe just remove UnknownMessageHandler - perhaps something else could be impl'd later (after this PR).

yellows8 commented on nx/include/switch/services/applet.h:
```diff
@@ -468,3 +468,21 @@ u32 appletGetPerformanceMode(void);
 Result appletSetFocusHandlingMode(AppletFocusHandlingMode mode);
+
+/**
+ * @brief Stops forwarding the input to the foreground app, works only in the Overlay applet context
```
*context.

yellows8 commented on nx/include/switch/services/hidsys.h:
```diff
@@ -0,0 +1,19 @@
+Result hidsysEnableAppletToGetInput(bool enable);
+
+/**
+* @brief Returns an event that fires when the home button is pressed, this will prevent the home menu from opening when the button is pressed. This event does not auto clear
```
*clear.

yellows8 commented on nx/source/services/hidsys.c:
```diff
@@ -0,0 +1,215 @@
+    struct {
+        u64 magic;
+        u64 cmdid;
+        bool permitInput;
```
u8

yellows8 commented on nx/source/services/hidsys.c:
```diff
@@ -0,0 +1,215 @@
+
+    raw->magic = SFCI_MAGIC;
+    raw->cmdid = 503;
+    raw->permitInput = enable;
```
enable!=0

yellows8 commented on nx/source/services/hidsys.c:
```diff
@@ -0,0 +1,215 @@
+    } *raw;
+
+    ipcSendPid(&c);
+    raw = serviceIpcPrepareHeader(&g_hidsysSrv,&c, sizeof(*raw));
```
`, &c`

yellows8 commented on nx/source/services/hidsys.c:
```diff
@@ -0,0 +1,215 @@
+    } *raw;
+
+    ipcSendPid(&c);
+    raw = serviceIpcPrepareHeader(&g_hidsysSrv,&c, sizeof(*raw));
```
see above

yellows8 commented on nx/source/services/applet.c:
```diff
@@ -2311,7 +2317,47 @@ bool appletMainLoop(void) {
     return true;
 }
+
+Result appletBeginToWatchShortHomeButtonMessage(void) {
```
Move these 3 funcs after appletQueryApplicationPlayStatistics.

yellows8 commented on nx/source/services/applet.c:
```diff
@@ -2311,7 +2317,47 @@ bool appletMainLoop(void) {
+}
+
+Result appletHomeButtonReaderLockAccessorGetEvent(Event *out_event)
+{
```
Move { to previous line, likewise elsewhere.

yellows8 commented on nx/source/services/applet.c:
```diff
@@ -2311,6 +2351,7 @@ bool appletMainLoop(void) {
 
             appletCallHook(AppletHookType_OnPerformanceMode);
         break;
+     
```
revert

yellows8 commented:
Can you return the Events directly to the user, without saving them in state?

yellows8 commented:
Yes, same thing elsewhere in libnx.

exelix11 commented:
Then the user will have to manually close them before exiting hidsys idk what would happen otherwise, is that OK ?

yellows8 commented:
Likewise for applet Events.

yellows8 commented on nx/source/services/hidsys.c:
```diff
@@ -0,0 +1,183 @@
+#include "result.h"
+#include "arm/atomics.h"
+#include "kernel/ipc.h"
+#include "kernel/shmem.h"
```
Remove this include and the next line.

yellows8 commented on nx/source/services/hidsys.c:
```diff
@@ -0,0 +1,183 @@
+    return _hidsysGetEvent(event_out, 121, false);
+}
+
+Result hidsysActivateHomeButton(void) {
```
Add these to the .h.

yellows8 commented on nx/include/switch/services/applet.h:
```diff
@@ -468,3 +468,21 @@ u32 appletGetPerformanceMode(void);
 AppletFocusState appletGetFocusState(void);
 
 Result appletSetFocusHandlingMode(AppletFocusHandlingMode mode);
+
```
Move these before appletPushToGeneralChannel in the .h.

fincs commented on nx/include/switch/services/applet.h:
```diff
@@ -467,4 +485,4 @@ u8  appletGetOperationMode(void);
 AppletFocusState appletGetFocusState(void);
 
-Result appletSetFocusHandlingMode(AppletFocusHandlingMode mode);
+Result appletSetFocusHandlingMode(AppletFocusHandlingMode mode);
```
Don't remove trailing newline

fincs commented on nx/include/switch/services/hidsys.h:
```diff
@@ -0,0 +1,23 @@
+
+Result hidsysActivateHomeButton(void);
+Result hidsysActivateSleepButton(void);
+Result hidsysActivateCaptureButton(void);
```
Add newline at EOF

fincs commented on nx/source/services/hidsys.c:
```diff
@@ -0,0 +1,181 @@
+
+Result hidsysActivateCaptureButton(void) {
+    return _hidsysCmdWithResIdAndPid(151);
+}
```
Add newline at EOF
